packages/editor/src/components/entities-saved-states/entity-record-item.js

```diff
@@ -28,6 +28,30 @@ export default function EntityRecordItem( {
 		return parents[ parents.length - 1 ];
 	}, [] );
 
+	// Handle templates that might use default descriptive titles
+	const entityRecordTitle = useSelect(
+		( select ) => {
+			if (
+				! select( 'core/edit-site' ) ||
+				'postType' !== kind ||
+				'wp_template' !== name
+			) {
+				return title;
+			}
+
+			const template = select( 'core' ).getEditedEntityRecord(
+				kind,
+				name,
+				key
+			);
+			const { title: templateTitle } = select(
+				'core/edit-site'
```

Comment:
This feels really weird to me -- are we sure this package can rely on the edit site store being available all the time? I wonder if there are scenarios where it could be used when the store hasn't been registered

Reply:
There is an early return for `! select( 'core/edit-site' )` a few lines above. 🙂 

Anyhow, I'm not happy with this as well.
I would absolutely love to move the `getTemplateInfo` selector elsewhere shared (e.g. `core/editor`).
However, there is an inconsistency between edit-post and edit-site initialization that prevent this, as settings initialized in PHP are fed to `core/editor` in edit-post, and `core/edit-site` in edit-site.

My opinion is that we should reduce/remove all inconsistencies between the different editor inits, but this is not the place.

I should probably add a todo comment explaining what happens here.

Reply:
It creates a cyclic dependency between `edit-site` and `editor` and should be avoided. We have just landed API extensions to `@wordpress/data` (https://github.com/WordPress/gutenberg/pull/26655) that is going to prevent this type of reference by using direct store definition imports. Would it be possible to pass `entityRecordTitle` as part of the `record.title` prop passed to `EntityRecordItem` instead or something along those lines?

Reply:
This originates from 2 different entities quirks:
- `auto-draft` records have their [title wiped out](https://github.com/WordPress/gutenberg/blob/e0041be4029d0447cb5866d99388ad35345de73a/packages/core-data/src/actions.js#L90) when added to the state.
- `wp_template` (and parts) records titles [fall back to the slug](https://github.com/WordPress/gutenberg/blob/e0041be4029d0447cb5866d99388ad35345de73a/packages/core-data/src/entities.js#L137-L143) instead of the id.

What we want is for templates to use the default title whenever possible.
This normally would mean to create them with the default title (which is what we do), but then when the template becomes an entity, given it is an `auto-draft`, it loses the title.
This in turn requires us to have that `getTemplateInfo` selector which gives templates back their intended title (or handles the various fallbacks).

The `EntityRecordItem` is, as far as I can see, the only place that would need this "title fix", as it's not tied to any specific entity.
But its part of the `editor` package, which shouldn't need `edit-site`.
The post editor initializes its settings into `core/editor`, while the Site Editor into `core/edit-site`. I really don't like it, but it's how it is right now.

In one of my previous attempts, I tried adding an exception to the [entities actions](https://github.com/WordPress/gutenberg/blob/e0041be4029d0447cb5866d99388ad35345de73a/packages/core-data/src/actions.js#L90): if it's a `wp_template`, don't clear the title.
But that was a bit smelly, and eventually I decided to avoid introducing exceptions in such a key part of the entire app, without understanding all the possible consequences.

The only other option (as far as I can see) it's to handle it whenever needed, which... is where we are right now! 😄 

Looking at `EntityRecordItem`, I don't see a good way to pass the "generated" template title without introducing exceptions to the dirty entities, which is something I'd rather avoid.

On the other hand, I'm exploring what it would take to update the Site Editor init to register the default template types to the `core/editor` settings instead of the `core/edit-site`'s, removing its unnecessary dependency.

Reply:
I've pushed an experimental commit to demonstrate what I mean with "moving `defaultTemplateTypes` to `core/editor`: 827464c037526bbec1fa54f8a228da5105f1db9f

Specifically this is where it happens:
https://github.com/WordPress/gutenberg/pull/27038/commits/827464c037526bbec1fa54f8a228da5105f1db9f#diff-53f39da7a5f1f996f5359a2dd7b2747fb3d60a494523dd73dba946fab8e82592R115-R119

Reply:
> On the other hand, I'm exploring what it would take to update the Site Editor init to register the default template types to the core/editor settings instead of the core/edit-site's, removing its unnecessary dependency.

It's not optimal and it's hacky. But I don't think we have a better option due to the way of how different editors initialized. ✅